Accessibility : manage aria-describedby attribute for feedback state

Lausselloic opened the pull request:
Provide a way to associate input fields with the error feedback for screen-reader

Accessibility improvement, manage aria-describedby attribute when valid / invalid state is update for an input field with feedback state

Open for discussion, don't hesitate to adapat it.

Lausselloic committed:
Accessibility improvement, manage aria-describedby attribute when valid / invalid state is update for an input field with feedback state

Lausselloic committed:
Merge branch 'v4-dev' into v4-dev

XhmikosR reviewed (changes requested):
This whole snippet should match our JS style, with spaces after keywords, no semicolons etc.

patrickhlauke commented:
Surprised/annoyed that I missed this new custom validation stuff (should really sit down and go over the whole of the docs again to see if there's anything else that's new but problematic).

In principle yes, agree that `aria-describedby` is needed here. Maybe also programmatically add/remove `aria-invalid="true"` ?

Johann-S commented on docs/4.0/components/forms.md:
```diff
@@ -765,13 +765,37 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
+      }
+
+      // Listen to input update, and check again the validity
+      inputs[i].addEventListener('input', function() {
```
so much event listeners added in case of `input` event triggered, not a good practice in terms of perf :worried: 
this should be done before the first call of `updateAriaDescribedAttribute`
with : 
```javascript
inputs.forEach(function (input) {
  input.addEventListener('input', function () {
    updateAriaDescribedAttribute([this]);
  })
})
```
and no more listeners added when `updateAriaDescribedAttribute` is called

Lausselloic committed:
fix JS style, and eventlistener management

Lausselloic committed:
Merge branch 'v4-dev' into v4-dev

Johann-S commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,43 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
-    var form = document.getElementById('needs-validation');
-    form.addEventListener('submit', function(event) {
+  window.addEventListener('load', () => {
+    const form = document.getElementById('needs-validation')
```
Not sure if we should use ES6 in our examples :thinking:

Lausselloic committed:
add management of aria-invalid state.

As some Browser/At vocalize invalid field on the first input, need to set aria-invalid to false by default. And after that, change the value according to the state

Lausselloic committed:
remove ES6 syntax as example are JS

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -700,7 +700,7 @@ While Bootstrap will apply these styles in all browsers, Internet Explorer 11 an
 
 {% callout warning %}
-We **highly recommend** custom validation styles as native browser defaults are not announced to screen readers.
+We **highly recommend** custom validation styles as native browser defaults are not announced to screen readers. Please ensure to manage `aria-describedby` attribute on `input` element according to the validation state.
```
i'd add a comma after "validation styles" to make it easier to read.

patrickhlauke replied:
"ensure to" probably better changed to "ensure you" (my preference). Otherwise "make sure to manage" sounds more natural. also good place to include the `aria-invalid` thing in the prose, e.g. "Please ensure you manage the `aria-invalid` and `aria-describedby` attributes on the `input` element..."

Lausselloic committed:
fix accessibility warning following @patrickhlauke review

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,45 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
+  // Works only if the invalid-feedback element have an ID
+  function updateAriaDescribedAttribute(inputs) {
+    for (var i = 0; i < inputs.length; i++) {
+      // Check element validity, and set aria-described attribute
```
`and set aria-describedby and aria-invalid attributes`

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,45 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
-        event.stopPropagation();
+        event.preventDefault()
+        event.stopPropagation()
+        // Check each form elements validity and add an aria-describedby to associate properly feedback with input
```
`Check each form element's validity and set appropriate aria-invalid / aria-describedby attributes`

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,45 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
+        event.stopPropagation()
+        // Check each form elements validity and add an aria-describedby to associate properly feedback with input
+        var inputs = form.getElementsByTagName('input')
+        updateAriaDescribedAttribute(inputs)
```
change function name as per above comment

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,45 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
+        // Listen to input update, and check again the validity
+        for (var i = 0; i < inputs.length; i++) {
+          inputs[i].addEventListener('input', function () {
+            updateAriaDescribedAttribute([this])
```
ditto

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -765,17 +765,45 @@ When attempting to submit, you'll see the `:invalid` and `:valid` styles applied
+        var inputs = form.getElementsByTagName('input')
+        updateAriaDescribedAttribute(inputs)
+
+        // Listen to input update, and check again the validity
```
`and check validity again`

patrickhlauke commented on docs/4.0/components/forms.md:
```diff
@@ -700,7 +700,7 @@ While Bootstrap will apply these styles in all browsers, Internet Explorer 11 an
 
 {% callout warning %}
-We **highly recommend** custom validation styles as native browser defaults are not announced to screen readers.
+We **highly recommend** custom validation styles, as native browser defaults are not announced to screen readers. Please ensure you manage the `aria-invalid` and `aria-describedby` attribute on `input` element according to the validation state.
```
may actually be good to soften this warning a bit, as some browsers now seem to work ok. suggest `We recommend using custom validation styles, as native browser validation errors are not consistently announced by all browser and screen reader combinations.`

Lausselloic replied:
yes, will finish to test that. Actually doesn't work properly with FFESR52 and NVDA 2017 even with the aria-invalid false, the browser still announce field as invalid when the user first get in the input field

Lausselloic replied:
Wording updated. But also a bug filled to Mozilla. https://bugzilla.mozilla.org/show_bug.cgi?id=1413193
By my tests, aria-invalid status works well on Chrome and IE11, but don't on Firefox with NVDA or Jaws

patrickhlauke reviewed (changes requested):
Sorry, few more changes I just spotted

Lausselloic committed:
update function naming as it update 2 aria attributes

XhmikosR committed:
Update forms.md

Lausselloic committed:
update warning message.

Also update attach event listener to input field, remove it out of the loop, otherwise event will be attached at each form submission. By the way add a check in updateValidityAriaAttribute if forms have already been submitted or not to prevent poping error message to screen-reader before the first submission

Lausselloic committed:
Merge branch 'v4-dev' into v4-dev

mdo commented:
This is too much JS to casually put into a code snippet. Should this be a plugin of some sort instead?

danijelGombac commented:
I saw this PR and have a question about aria-describedby. What if aria-describedby is already used because form-text. Then should be changed to invalid-feedback id when is invalid and back when is valid?

patrickhlauke commented:
@mdo 

> This is too much JS to casually put into a code snippet. Should this be a plugin of some sort instead?

generalising this into a plugin of sorts would probably only work if we then also mandate exactly how authors code/add the extra validation/error message, or make the plugin super-configurable in terms of being able to say on each input how/where any error message is in the document in order to then add/remove the `aria-describedby` reference to it.

I'm almost tempted to say this goes beyond what Bootstrap itself would/should tackle directly...

But agree that it's quite a chunk of JS just thrown in there in passing.

patrickhlauke commented:
@danijelGombac 

>I saw this PR and have a question about aria-describedby. What if aria-describedby is already used because help text (form-text). Then should be changed to invalid-feedback id when is invalid and back when is valid?

`aria-describedby` takes one or more IDs (space-separated list of IDs), so if you're already using `aria-describedby` to point to additional info about the form field, then when there's a related error/validation message, you'd want BOTH the id for the additional info AND the id of the error message

XhmikosR committed:
Update forms.md

XhmikosR committed:
Merge branch 'v4-dev' into v4-dev

tmorehouse commented:
@patrickhlauke on a related note about `aria-describedby`.  Popovers and Tooltips add `aria-describedby` to the trigger element when the tooltip or popover is opened (and removes it when closed), but they don't respect any existing `aria-describedby` ids that may already be on the trigger elements (they overwrite and completely remove).

patrickhlauke commented:
@tmorehouse could you file separate issue(s) for those?

tmorehouse commented:
@patrickhlauke issue #24640 has been created.

Lausselloic committed:
Merge branch 'v4-dev' into v4-dev

XhmikosR committed:
Merge branch 'v4-dev' into v4-dev

XhmikosR commented:
@patrickhlauke: LGTY?

XhmikosR commented:
Not sure what to do with this. It has conflicts.

XhmikosR commented:
Closing since they are conflicts. Feel free to make a new PR referencing this one.
